### PR TITLE
Add Docker Compose example

### LINKED ISSUE #31: Docker Support
Is a docker (docker compose) support planned ?

### PR DESCRIPTION
**Pull request by easybe**
This demonstrates how to deploy the service in production using the
Docker image together with NGINX and PostgreSQL.

Run with:

    docker compose -f docker/demo/docker-compose.yml up

The service can then be accessed under: https://localhost

Closes #31

### CODE REVIEW

File: docker/Dockerfile [hunk 1]
```diff
@@ -16,7 +16,7 @@ EXPOSE 60053
 
 USER goosebit
 
-ENV GUNICORN_CMD_ARGS="--workers 2 --enable-stdio-inheritance"
+ENV GUNICORN_CMD_ARGS="--config gunicorn_config.py --workers 2 --enable-stdio-inheritance"
```
Comment (easybe) [inline docker/Dockerfile:19]: This needs to be moved to the Compose file.

**Review by b-rowan — APPROVED** [review]
Looks good to me, will do some testing.

Only concern is the hard coding of variables (like GOOSEBIT_SECRET), but my understanding is we want people to change this anyway so that's fine.

**Comment by b-rowan** [comment]
Trying to upload software is failing for me, with the following error -

```
nginx-1     | 2024/09/11 19:27:06 [warn] 24#24: *4 a client request body is buffered to a temporary file /var/cache/nginx/client_temp/0000000023, client: 172.18.0.1, server: _, request: "POST /ui/bff/software HTTP/2.0", host: "0.0.0.0", referrer: "https://0.0.0.0/ui/software"
```

**Comment by easybe** [comment]
> Looks good to me, will do some testing.
> 
> Only concern is the hard coding of variables (like GOOSEBIT_SECRET), but my understanding is we want people to change this anyway so that's fine.

This is just an example and of course, should not be used as-is. With Docker Compose secrets can be put in a `.env` file. How secrets are handled heavily depends on the software you are using for deployment.

**Comment by easybe** [comment]
> Trying to upload software is failing for me, with the following error -
> 
> ```
> nginx-1     | 2024/09/11 19:27:06 [warn] 24#24: *4 a client request body is buffered to a temporary file /var/cache/nginx/client_temp/0000000023, client: 172.18.0.1, server: _, request: "POST /ui/bff/software HTTP/2.0", host: "0.0.0.0", referrer: "https://0.0.0.0/ui/software"
> ```

It's a permissions issue. We should fix that in the image. Will do that together with the other pending fix.

**Comment by easybe** [comment]
@b-rowan how about a v0.2.4? The latest image on Docker Hub is still broken, which is a problem if someone wants to try it out…

**Comment by b-rowan** [comment]
On it 🫡

**Review by b-rowan — APPROVED** [review]
LGTM 🚀

**Comment by embetrix** [comment]
I get **"Software swu header cannot be parsed"** error message when using the docker compose and upload an SWU file.

**Comment by b-rowan** [comment]
> I get **"Software swu header cannot be parsed"** error message when using the docker compose and upload an SWU file.

What format are you using for versioning?  That's likely where that's failing, right now goosebit expects a semantic version.

**Comment by easybe** [comment]
It could also be a permissions problem (`/artifacts`). Make sure you pull the latest image and re-create the volumes:
```
docker compose down --volumes
docker compose up
```

**Comment by embetrix** [comment]
my swu package is : **my-image-qtdemo-update-stm32mp157f-dk2_0.6.1.swu** are there specific format it should be followed ?

the volumes are clean, I just pulled today.

**Comment by easybe** [comment]
Unfortunately, logging needs some fixing. Could you try again with
```
image: upstreamdata/goosebit:wip
```
and post the log output?

**Comment by embetrix** [comment]
Ah, didn't see this information in the docs. I use by the way encrypted **sw-description** : 
https://github.com/sbabic/swupdate/blob/master/Kconfig#L494

IMHO Goosebit should not have this restriction :-(

**Comment by easybe** [comment]
> Ah, didn't see this information in the docs. I use by the way encrypted **sw-description** : https://github.com/sbabic/swupdate/blob/master/Kconfig#L494
> 
> IMHO Goosebit should not have this restriction :-(

Keep in mind that gooseBit is still under heavy development and not feature complete/bug free. If you have ideas/suggestions on how to improve it, feel free to create an issue.

**Pull request merged**